### PR TITLE
fixes #50 - generates pvcell coeffs from iec 61853

### LINKED ISSUE #50: must know pvcell coefficients a priori
if you don't know the pvcell coefficients you can't use pvmismatch - kind of makes pvmismatch useless :frowning_face:

### PR DESCRIPTION
**Pull request by mikofski**
* calculate iec 61853 from sandia perfmod

Signed-off-by: Mark Mikofski <mark.mikofski@sunpowercorp.com>

### CODE REVIEW

**Comment by mikofski** [comment]
@bmeyers this is ready for review for like a release candidate release. Can you take a quick look and let me know if you think it's okay to merge and release as v3.1rc1? Don't worry about the ugly jacobian expressions, they were generated by sympy, but I didn't simplify them at all, takes too much time and the calculation savings is in &mu;s. Their validity is checked in the tests, so if you're inclined you can validate the sympy expressions.

I won't work any more on this for awhile, but it is good enough to use with PVMismatch the way that PVMismatch is currently implemented, _IE_: with 2-diode model and no irradiance modification of either `Isat2` or `Rsh`.

Unfortunately the errors of off-STC conditions vary widely between module types. This will be resolved by switching to either CEC or PVSyst model or by adding modifiers to either Rsh, Isat2 or both.

The `gen_coeff` methods can be used with arrays of input if the optimizer method is set to `'lm'`, or input at any `(irr, tc)` condition using the default `'hbrj'` solver. Generating coefficients over the entire IEC61853 matrix do confirm that `Isat1` and `Rs` are constant for all `(irr, tc)`, but `Isat2` and `Rsh` vary with irradiance.

The next time we visit this, we should look into resolving the irradiance dependence of `Isat2` and `Rsh`.

**Comment by bmeyers** [comment]
@mikofski taking a look now. looks like a lot of work!

File: pvmismatch/contrib/gen_coeffs/diode.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    vd_isat_growth = -vd * isat_growth  # combination parameter
+    # diode current
+    id_ = np.atleast_1d(isat * expfact)
+    # derivatives
+    d_isat = np.atleast_1d(expfact)  # df w.r.t. isat
+    d_vd = np.atleast_1d(isat_growth / vact)  # df w.r.t. vd
+    d_m = np.atleast_1d(vd_isat_growth / (m ** 2.0 * vt))  # df w.r.t. m
+    d_vt = np.atleast_1d(vd_isat_growth / (m * vt ** 2.0))  # df w.r.t. vt
+    jac = np.array([d_isat, d_vd, d_m, d_vt])  # jacobian
+    return id_, jac
```
Comment (bmeyers) [inline pvmismatch/contrib/gen_coeffs/diode.py]: :thumbsup: math checks out

**Comment by BenBourne** [comment]
Mark, have we done any testing with measured 61853 data to confirm that the generated coefs are consistent with the test data?

-Ben Bourne
SunPower Corporation
650-208-6957 (c)
510-260-8388 (o)

Sent from my mobile phone

On Mar 23, 2017, at 10:28 AM, Mark Mikofski <notifications@github.com<mailto:notifications@github.com>> wrote:


@bmeyers<https://github.com/bmeyers> this is ready for review for like a release candidate release. Can you take a quick look and let me know if you think it's okay to merge and release as v3.1rc1? Don't worry about the ugly jacobian expressions, they were generated by sympy, but I didn't simplify them at all, takes too much time and the calculation savings is in μs. Their validity is checked in the tests, so if you're inclined you can validate the sympy expressions.

I won't work any more on this for awhile, but it is good enough to use with PVMismatch the way that PVMismatch is currently implemented, IE: with 2-diode model and no irradiance modification of either Isat2 or Rsh.

Unfortunately the errors of off-STC conditions vary widely between module types. This will be resolved by switching to either CEC or PVSyst model or by adding modifiers to either Rsh, Isat2 or both.

The gen_coeff methods can be used with arrays of input if the optimizer method is set to 'lm', or input at any (irr, tc) condition using the default 'hbrj' solver. Generating coefficients over the entire IEC61853 matrix do confirm that Isat1 and Rs are constant for all (irr, tc), but Isat2 and Rsh vary with irradiance.

The next time we visit this, we should look into resolving the irradiance dependence of Isat2 and Rsh.

**Comment by mikofski** [comment]
@BenBourne unfortunately, the generated 2-diode coefficients have two issues:

1. The 2-diode model as implemented (without irradiance modifiers for Rsh and Isat2) do not fit all of the data well. In particular there are errors at low temperature and high temperature. The errors vary from module to module. EG for P series the errors were smaller than for E series.
2. The 2-diode model as implemented fits better when only STC values are used. When all IEC61853 measurements are used, the variaitions in Rsh and Isat2 are too large to be fit by a common set of parameters for the entire range of IEC61853. In order to fit the entire range we will have to either:
    * use irradiance modifiers with either Rsh, Isat2 or both
    * switch to either CEC or PVSyst model

This is as good as we can get it for right now. When we have time to revisit this, and improve the models to reduce the error. For now I think this PR satisfies the need to provide PVMismatch parameters at STC for any module in the SAPM database.

# P-series (fit with only STC measurments)
pvmismatch_gen_p17-345w_coeffs

# E-series (fit with only STC measurments)
pvmismatch_gen_coeffs

File: pvmismatch/contrib/gen_coeffs/diode.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+
+def isat_t(tc, isat0):
+    tck = tc + 273.15
+    t0k = T0 + 273.15
+    tstar = tck ** 3.0 / t0k ** 3.0
+    inv_delta_t = 1.0 / t0k - 1.0 / tck  # [1/K]
+    exp_tstar = np.exp(EG * QE / KB * inv_delta_t)
+    isat_t = isat0 * tstar * exp_tstar
+    return isat_t
```
Comment (bmeyers) [inline pvmismatch/contrib/gen_coeffs/diode.py]: I'm not familiar with this equation. Do you mind sending me a reference so I can get up to speed?

**Comment by bmeyers** [comment]
Excellent work, Mark!

**Comment by mikofski** [comment]
For cubic dependence of Isat1 see
* [PVPMC description of CEC model](https://pvpmc.sandia.gov/modeling-steps/2-dc-module-iv/diode-equivalent-circuit-models/de-soto-5-parameter-model/)
* [PV CD-ROM effect of temperature] (http://www.pveducation.org/pvcdrom/effect-of-temperature)

Also in tests with data for PVLife the first diode in two diode model fits this relation very close. In only one diode model doesn't fit as well

**Pull request merged**